sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/SynchronousMessageSubscriber.java

```diff
@@ -5,114 +5,220 @@
      */
     @Override
-    protected void hookOnNext(ServiceBusReceivedMessageContext value) {
-        work.next(value);
+    protected void hookOnNext(ServiceBusReceivedMessageContext message) {
+        boolean delivered = false;
+
+        // Boundary condition(timeout case): if timeout has happened, get next available work.
+        if (currentWork == null) {
+            // Now see if there is more queued up work
+            currentWork = getNextWorkAndRequest();
+            logger.verbose("No current work, Picked up next receive request.");
+        }
+
+        if (currentWork != null) {
+            currentWork.next(message);
+            delivered = true;
+            remaining.decrementAndGet();
+            // Check if we have delivered all the messages to current work.
+            if (currentWork.isTerminal()) {
+                currentWork.complete();
+                if (timeoutOperation != null && !timeoutOperation.isDisposed()) {
+                    timeoutOperation.dispose();
+                }
+
+                // Now see if there is more queued up work
+                currentWork = getNextWorkAndRequest();
+                logger.verbose("Current work is terminal, Picked up next receive request.");
+            }
+        }
 
-        if (work.isTerminal()) {
-            logger.info("[{}] Completed. Closing Flux and cancelling subscription.", work.getId());
-            dispose();
+        if (currentWork == null) {
+            if (wip.decrementAndGet() != 0) {
+                logger.warning("There is another worker in drainLoop. But there should only be 1 worker.");
+            }
+            // If message is not delivered to downstream, we will buffer it.
+            if (!delivered) {
+                bufferMessages.add(message);
+            }
         }
     }
 
-    @Override
-    protected void hookOnComplete() {
-        logger.info("[{}] Completed. No events to listen to.", work.getId());
-        dispose();
+    /**
+     *
+     * @param requested credits for current {@link SynchronousReceiveWork}.
+     */
+    private void requestCredits(long requested) {
+        long creditToAdd = requested - (remaining.get() + bufferMessages.size());
+        if (creditToAdd > 0) {
+            remaining.addAndGet(creditToAdd);
+            subscription.request(creditToAdd);
+        }
+    }
+
+    void queueWork(SynchronousReceiveWork work) {
+
+        logger.info("[{}] Pending: {}, Scheduling receive timeout task '{}'.", work.getId(), work.getNumberOfEvents(),
+            work.getTimeout());
+        workQueue.add(work);
+        drain();
     }
 
+    private void drain() {
+        if (workQueue.size() == 0) {
+            return;
+        }
+
+        // If someone is already in this loop, then we are already clearing the queue.
+        if (!wip.compareAndSet(0, 1)) {
+            return;
+        }
+
+        // Drain queue..
+        drainQueue();
+
+    }
+
+
+    private void drainQueue() {
```

Comment:
1. There are 4 places in which you set `currentWork` and these can happen asynchronously. It's possible that the work you're referencing here is one that has been cancelled or timed out. Consider a `currentWorkLock` that synchronises read-write to this variable and the associated timeout operation.
2. Looking at other implementations of subscribers and your current implementation, I would put most of the heavy lifting in this method. (ie. checking to see if we have current work, updating it if need be, setting/requesting more work). As it currently is, there are several ways to emit next items and the currentWork is controlled in many places which will be a pain to debug when this fails.

Reply:
You can look at any of the operators for inspiration.
https://github.com/reactor/reactor-core/blob/master/reactor-core/src/main/java/reactor/core/publisher/FluxCreate.java#L716